### PR TITLE
Lots of fixes

### PR DESCRIPTION
**Pull request by nyalldawson**
All related to the project from https://issues.qgis.org/issues/14752 - this was firing up heaps of coordinate transform errors and exceptions, and lots of geos warnings from different code paths.

Let's freakin fix them all and make this project absolutely fly....

### CODE REVIEW

File: src/core/geometry/qgslinestring.cpp [hunk 1]
```diff
@@ -1329,3 +1344,49 @@ bool QgsLineString::convertTo( QgsWkbTypes::Type type )
     return QgsCurve::convertTo( type );
   }
```
Comment (m-kuhn) [inline src/core/geometry/qgslinestring.cpp]: Else?

**Comment by nirvn** [comment]
Lots of joy.

File: src/core/geometry/qgscircularstring.cpp [hunk 2]
```diff
@@ -546,6 +546,52 @@ double QgsCircularString::yAt( int index ) const
...
+  double *destM = srcM;
+  double *destZ = srcZ;
+
+  int filteredPoints = 0;
+  for ( int i = 0; i < size; ++i )
+  {
+    double x = *srcX++;
+    double y = *srcY++;
+    double z = hasZ ? *srcZ++ : std::numeric_limits<double>::quiet_NaN();
+    double m = hasM ? *srcM++ : std::numeric_limits<double>::quiet_NaN();
```
Comment (m-kuhn) [inline src/core/geometry/qgscircularstring.cpp]: Just a fun idea, could use a `signalling_NaN` and then check if a floating point exception has been raised to inform about filters that do not check for existence of z/m coordinates. :P

**Comment by wonder-sk** [comment]
Great work!

There are few inconsistencies in the filtering algorithm:
- interior rings of a polygon are removed if they collapse, but the exterior ring stays. Linestrings also are not turned into empty linestring if there is just one point left. Not sure if we should try to remove collapsed interior rings - there are other ways how the geometries can get invalid that we do not handle so maybe better to leave the decision on the caller what to do next?
- I guess multi-point geometries won't get filtered?

**Comment by nyalldawson** [comment]
@wonder-sk 

Thanks for the review!

> interior rings of a polygon are removed if they collapse, but the exterior ring stays. Linestrings also are not turned into empty linestring if there is just one point left. Not sure if we should try to remove collapsed interior rings - there are other ways how the geometries can get invalid that we do not handle so maybe better to leave the decision on the caller what to do next?

Ok. I'll make a separate "removeInvalidRings" function which does this step.
 
> I guess multi-point geometries won't get filtered?

Good catch -- I'll fix!

**Pull request merged**